### PR TITLE
Extend status report element

### PR DESCRIPTION
**Pull request by devanlooches**
Right now I added the full error inside of a stretchable and scrollable element on a right side bar.
What else would you like me to include in the status update element and also why did you want me to make it hold a vector?

### CODE REVIEW

**Review by hannobraun — COMMENTED** [review]
Thank you, @devanlooches!

Do you think this could be placed into a corner (lower-right? doesn't matter, maybe), and made more broad than high? Then there would be fewer line breaks in the error message.

> What else would you like me to include in the status update element

No more ideas right now, but now that we have it, it's easy to add more when we need to :+1: 


> and also why did you want me to make it hold a vector?

I figured then we could display the last few status messages, and not just the last one.

**Comment by devanlooches** [comment]
> Do you think this could be placed into a corner (lower-right? doesn't matter, maybe), and made more broad than high? Then there would be fewer line breaks in the error message.

We could place it there but the element is supposed to auto resize itself to the contents (I don't know why it breaks the error message). But it is also stretchable so you can extend it horizontally by dragging the edge outwards.

> I figured then we could display the last few status messages, and not just the last one.

How many status messages do you think should be kept before clearing it?

**Comment by devanlooches** [comment]
I could make the element completely draggable that way the user can put wherever they want.

**Review by hannobraun — COMMENTED** [review]
Thank you for the additional commits, @devanlooches, looks good!

> We could place it there but the element is supposed to auto resize itself to the contents (I don't know why it breaks the error message). But it is also stretchable so you can extend it horizontally by dragging the edge outwards.

Hmm, when I grab it at its edge, I can move it, but not resize it. After I stop pressing the left mouse button, it resets to its original position.

> I could make the element completely draggable that way the user can put wherever they want.

That might be ideal! Is that easy to do?

> How many status messages do you think should be kept before clearing it?

I don't know. Maybe 5?

It doesn't really matter. Once the mechanism is there to store/display multiple messages and limit them to a certain number, it's easy to adjust that number later, as we gain more experience with it.

**Comment by devanlooches** [comment]
Cool I will implement all of this shortly its a pretty quick fix.

**Comment by hannobraun** [comment]
Nice, thank you!

**Comment by devanlooches** [comment]
Here is the fix. Unfortunately I'm having trouble putting the window on the bottom right by default. If anyone could help me with that it would be great.

**Comment by devanlooches** [comment]
One other thing I was thinking is it might be a good idea to keep the 5 status messages unless there is a compiler error in which case it might be a good idea to clear them in order to save space for the compiler error message (And also not show it more than once).

**Comment by hannobraun** [comment]
Thank you, @devanlooches, I'll take a look later!

File: crates/fj-interop/src/status_report.rs [hunk 1]
```diff
@@ -2,25 +2,31 @@
...
-        }
+        Self { status: Vec::new() }
     }
 
     /// Update the status
     pub fn update_status(&mut self, status: &str) {
-        self.status = status.to_string();
+        if self.status.len() >= 5 {
+            self.clear_status();
+        }
```
Comment (hannobraun) [inline crates/fj-interop/src/status_report.rs:18]: I think this implementation is a bit weird, as it results in lots of statuses being removed at once. Might be better to just remove the oldest status, until only 5 are left. Maybe [`VecDeque`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html) is a better type for this than `Vec`, as it provides convenient `push_back`/`pop_front` methods.

But that's something that can be fixed later.

**Review by hannobraun — APPROVED** [review]
Thanks for the latest changes, @devanlooches! And sorry for the delay; I was out and about for much of the weekend.

> Unfortunately I'm having trouble putting the window on the bottom right by default (and still having it able to be dragged). If anyone could help me with that it would be great.

Unfortunately I don't know either (I don't have much experience with egui). We can always improve it later! (And maybe the lower-right corner isn't even the best location.)

> One other thing I was thinking is it might be a good idea to keep the 5 status messages unless there is a compiler error in which case it might be a good idea to clear them in order to save space for the compiler error message (And also not show it more than once).

That sounds good!

I think it makes sense to distinguish between different kinds of status updates (info, error, etc.) and make decision based on what types we currently have.

---

I'm merging this now! There's still room for improvement, of course, but this is already much better than what we currently have. Thanks you again, @devanlooches!

File: crates/fj-interop/src/status_report.rs (hunk 1 above)
Comment (devanlooches) [reply crates/fj-interop/src/status_report.rs:18]: No problem! Thanks for the feedback! Your right, VecDeque does look convenient. I'll look into it.

**Pull request merged**